lib/babe/epoch_test.go

```diff
@@ -0,0 +1,201 @@
+	mockEpochState1.EXPECT().SetFirstSlot(gomock.Eq(uint64(1))).Return(nil)
+
+	testBabeSecondaryPlainPreDigest := types.BabeSecondaryPlainPreDigest{
+		SlotNumber: 1,
+	}
+
+	encDigest := newEncodedBabeDigest(t, testBabeSecondaryPlainPreDigest)
+	header := newTestHeader(t, *types.NewBABEPreRuntimeDigest(encDigest))
+
+	block := &types.Block{
+		Header: *header,
+	}
+
+	mockBlockState.EXPECT().GetBlockByNumber(gomock.Eq(big.NewInt(1))).Return(block, nil)
+	mockBlockState.EXPECT().GetBlockByNumber(gomock.Eq(big.NewInt(1))).Return(block, nil)
+
+	bs0 := &Service{
+		epochState: mockEpochState0,
+		blockState: mockBlockState,
+	}
+
+	bs1 := &Service{
+		epochState: mockEpochState1,
+		blockState: mockBlockState,
+	}
+
+	cases := []struct {
+		name    string
+		service *Service
+	}{
+		{
+			name:    "should not set first slot, as it's already set correctly",
+			service: bs0,
+		},
+		{
+			name:    "should update first slot, as it's set incorrectly",
+			service: bs1,
+		},
+	}
+
+	for _, tc := range cases {
+		err := tc.service.checkAndSetFirstSlot()
+		require.NoError(t, err)
+	}
+}
+
+func TestBabeService_getEpochDataAndStartSlot(t *testing.T) {
+	ctrl := gomock.NewController(t)
+	mockBlockState := NewMockBlockState(ctrl)
+	mockEpochState0 := NewMockEpochState(ctrl)
+	mockEpochState1 := NewMockEpochState(ctrl)
+	mockEpochState2 := NewMockEpochState(ctrl)
+
+	mockEpochState0.EXPECT().GetStartSlotForEpoch(gomock.Eq(uint64(0))).Return(uint64(1), nil)
+	mockEpochState1.EXPECT().GetStartSlotForEpoch(gomock.Eq(uint64(1))).Return(uint64(201), nil)
+	mockEpochState2.EXPECT().GetStartSlotForEpoch(gomock.Eq(uint64(1))).Return(uint64(201), nil)
+
+	mockEpochState1.EXPECT().HasEpochData(gomock.Eq(uint64(1))).Return(true, nil)
+	mockEpochState2.EXPECT().HasEpochData(gomock.Eq(uint64(1))).Return(true, nil)
+
+	kp := keyring.Alice().(*sr25519.Keypair)
+	authority := types.NewAuthority(kp.Public(), uint64(1))
+	testEpochData := &types.EpochData{
+		Randomness:  [32]byte{1},
+		Authorities: []types.Authority{*authority},
+	}
+
+	mockEpochState1.EXPECT().GetEpochData(gomock.Eq(uint64(1))).Return(testEpochData, nil)
+	mockEpochState2.EXPECT().GetEpochData(gomock.Eq(uint64(1))).Return(testEpochData, nil)
+
+	mockEpochState1.EXPECT().HasConfigData(gomock.Eq(uint64(1))).Return(true, nil)
+	mockEpochState2.EXPECT().HasConfigData(gomock.Eq(uint64(1))).Return(false, nil)
+
+	testConfigData := &types.ConfigData{
+		C1: 1,
+		C2: 1,
+	}
+
+	mockEpochState1.EXPECT().GetConfigData(gomock.Eq(uint64(1))).Return(testConfigData, nil)
+
+	testLatestConfigData := &types.ConfigData{
+		C1: 1,
+		C2: 2,
+	}
+
+	mockEpochState2.EXPECT().GetLatestConfigData().Return(testLatestConfigData, nil)
+
+	testEpochDataEpoch0 := &types.EpochData{
+		Randomness:  [32]byte{9},
+		Authorities: []types.Authority{*authority},
+	}
+
+	mockEpochState0.EXPECT().GetLatestEpochData().Return(testEpochDataEpoch0, nil)
+	mockEpochState0.EXPECT().GetLatestConfigData().Return(testConfigData, nil)
```

Comment:
Perhaps you could configure your mocks according to test cases?

Like https://github.com/qdm12/gluetun/blob/55e609cbf43c76570c7831c5122ab121fd5c2382/internal/wireguard/rule_test.go#L84 for example.

There is also a reason where I believe having `NewController(t)` with the parent test might give some funny logs/behavior especially when running parallel subtests inside it.

Reply:
yeah that would be nice, I'll work on that. gotta work on making my tests cleaner haha